modules/cudastereo/src/cuda/stereobm.cu

```diff
@@ -313,29 +360,90 @@ namespace cv { namespace cuda { namespace device
+                        batch_ssds[2] = CalcSSD<RADIUS>(col_ssd_cache + threadIdx.x, col_ssd + 2 * (BLOCK_W + 2 * RADIUS), X, cwidth);
+                        __syncthreads();
+                        batch_ssds[3] = CalcSSD<RADIUS>(col_ssd_cache + threadIdx.x, col_ssd + 3 * (BLOCK_W + 2 * RADIUS), X, cwidth);
+                        __syncthreads();
+                        batch_ssds[4] = CalcSSD<RADIUS>(col_ssd_cache + threadIdx.x, col_ssd + 4 * (BLOCK_W + 2 * RADIUS), X, cwidth);
+                        __syncthreads();
+                        batch_ssds[5] = CalcSSD<RADIUS>(col_ssd_cache + threadIdx.x, col_ssd + 5 * (BLOCK_W + 2 * RADIUS), X, cwidth);
+                        __syncthreads();
+                        batch_ssds[6] = CalcSSD<RADIUS>(col_ssd_cache + threadIdx.x, col_ssd + 6 * (BLOCK_W + 2 * RADIUS), X, cwidth);
+                        __syncthreads();
+                        batch_ssds[7] = CalcSSD<RADIUS>(col_ssd_cache + threadIdx.x, col_ssd + 7 * (BLOCK_W + 2 * RADIUS), X, cwidth);
+
+                        uint mssd = ::min(::min(::min(batch_ssds[0], batch_ssds[1]), ::min(batch_ssds[4], batch_ssds[5])),
+                                         ::min(::min(batch_ssds[2], batch_ssds[3]), ::min(batch_ssds[6], batch_ssds[7])));
+
+                        int bestIdx = 0;
+                        for (int i = 0; i < N_DISPARITIES; i++)
+                        {
+                            if (mssd == batch_ssds[i])
+                                bestIdx = i;
+                        }
 
                         // For threads that do not satisfy the if condition below("X < cwidth - RADIUS"), previously
-                        // computed "minSSD" value, which is the result of "MinSSD" function call, is not used at all.
+                        // computed "minSSD" value, which is the result of "CalcSSDVector" function call, is not used at all.
                         //
-                        // However, since the "MinSSD" function has "__syncthreads" call in its body, those threads
-                        // must also call "MinSSD" to avoid deadlock. (#13850)
+                        // However, since the "CalcSSDVector" function has "__syncthreads" call in its body, those threads
+                        // must also call "CalcSSDVector" to avoid deadlock. (#13850)
                         //
                         // From CUDA 9, using "__syncwarp" with proper mask value instead of using "__syncthreads"
                         // could be an option, but the shared memory access pattern does not allow this option,
                         // resulting in race condition. (Checked via "cuda-memcheck --tool racecheck")
 
                         if (X < cwidth - RADIUS)
                         {
-                            int idx = row * cminSSD_step;
-                            if (minSSD.x < minSSDImage[idx])
+                            unsigned int last_opt = line_ssd_tails[3*row + 0];
```

Comment:
could it be moved to a separate function to decrease duplicated code? see line 300 with row = 0

Reply:
Theoretically yes. I tried to extract the function and faced the following issues:
- The function has very large context that is re-used on caller side. I do not want to re-compute some values several times.
- There are some simplifications for `row=0`.
- - The code become even less obvious, because most of variables are not function local and should be read "inline" to the caller side.